feat: application side nav

ColinBuyck opened the pull request:
This PR addresses https://github.com/metrotranscom/doorway/issues/437
- [x] Addresses the issue in full
- [ ] Addresses only certain aspects of the issue

## Description

This PR utilizes the tab component so a user can filter the applications they would like to see. 

Note One: This implementation results in applications being re-fetched each time they select a new category. Based on the performance of this, I am wondering if there is a more reduced fetch call that only provides the required information.

Note Two: I made some assumptions on the mobile experience of this section, let me know what y'all think!

Note Three: I was looking into expanding the test coverage and was considering ctypress tests but know we have been cautious in adding more cypress. Additionally, seeing how limited my-applications.spec.ts is would it be worth it to expand it? Planning on expending integration test coverage in #4053 

## How Can This Be Tested/Reviewed?

This can be tested by signing into the public site and first testing the empty states of the different sections. If there are no applications at all, the message will be to browse more listings but if you have applied but none of the applications are in the selected state, it will be a custom message. Then apply to multiple listings. Lastly, change the status of those listings and observe the applications move to the corresponding section

## Author Checklist:

- [x] Added QA notes to the issue with applicable URLs
- [x] Reviewed in a desktop view
- [x] Reviewed in a mobile view
- [x] Reviewed considering accessibility
- [ ] Added tests covering the changes
- [ ] Made corresponding changes to the documentation
- [ ] Ran `yarn generate:client` and/or created a migration when required

## Review Process:

- Read and understand the issue
- Ensure the author has added QA notes
- Review the code itself from a style point of view
- Pull the changes down locally and test that the acceptance criteria is met
- Either (1) explicitly ask a clarifying question, (2) request changes, or (3) approve the PR, even if there are very small remaining changes, if you don't need to re-review after the updates

ColinBuyck commented on sites/public/src/components/account/StatusItem.tsx:
```diff
@@ -24,7 +24,7 @@ interface StatusItemProps {
 const StatusItem = (props: StatusItemProps) => {
   return (
-    <Card.Section className={accountStyles["account-card-applications-section"]}>
+    <Card.Section className={applicationsViewStyles["account-card-applications-section"]}>
```
Is it preferred to duplicate this class instead of this crossover @emilyjablonski ?

emilyjablonski replied:
This lgtm!

emilyjablonski commented on shared-helpers/src/locales/general.json:
```diff
@@ -10,10 +10,18 @@
   "account.createAccount": "Create Account",
   "account.errorFetchingApplications": "Error fetching applications",
   "account.haveAnAccount": "Already have an account?",
+  "account.allMyApplications": "All my applications",
```
Do you mind alphabetizing these? W lines selected, I use command + shift + p --> sort lines ascending in VSCode

ColinBuyck replied:
Learn something new everyday

ludtkemorgan replied:
One call out here, since this whole file was alphabetized we will have to be careful when pull this over to the jurisdictions. There is a chance that we might not catch if one of the moved keys is a different translation (this has happened in the past). And I'd also recommend that the alphabetizing is also run on those jurisdiction's repos so that any keys that are unique to that repo are also placed in the right place.

emilyjablonski commented on sites/public/src/components/account/ApplicationsView.module.scss:
```diff
@@ -0,0 +1,44 @@
+.applications-section-background {
+  background: var(--seeds-color-gray-300);
```
Where possible it's preferable to use semantic tokens for theming, for example this could be `--seeds-bg-color-canvas` which is defaulted in our theme to `--seeds-color-gray-300`

ColinBuyck replied:
Oh nice, I haven't noticed the canvas one before!

emilyjablonski commented on sites/public/src/components/account/ApplicationsView.module.scss:
```diff
@@ -0,0 +1,44 @@
+.applications-section-background {
+  background: var(--seeds-color-gray-300);
+  border-color: var(--seeds-color-gray-450);
```
`--seeds-border-color`

emilyjablonski commented:
Not sure if it matters, but in the lottery designs, the order of the last two tabs is switched, the design goes all --> lottery run --> applications closed --> accepting applications

emilyjablonski commented:
When I test this in the deploy preview, I relatively quickly and am getting the issue where the listing data stops appearing, and it looks like what is happening is that I'm getting throttled. We may need to update the threshold. I'm interested in separately moving forward w a smaller view, created a ticket [here](https://github.com/bloom-housing/bloom/issues/4240) 🏹

emilyjablonski commented on sites/public/src/components/account/StatusItemWrapper.tsx:
Ty!

emilyjablonski commented:
<img width="991" alt="Screenshot 2024-08-07 at 7 34 30 PM" src="https://github.com/user-attachments/assets/e74dce2c-37a7-4e6a-979f-573f12e095a7">
Seeing this button when I am already on that page felt confusing, we maybe just need to remove it from that view.

emilyjablonski reviewed (changes requested):
Some feedback above!

jaredcwhite commented on sites/public/src/components/account/ApplicationsView.tsx:
```diff
@@ -0,0 +1,230 @@
+      })
+  }, [applications, listLoading, listingsService])
+
+  const selectionHandler = (index: number) => {
```
FWIW, this definitely leads me to believe that what we really need is the to-be-developed Seeds Menu component. The Tabs component is intended to switch views already loaded on-page without changing routes. I'm not sure if the a11y on this quite makes sense.

ludtkemorgan replied:
To add on to this, do we need to fully change routes when switching tabs? In the past when I've worked on a page like this I used a context to store the data so when switching between tabs I just grab the appropriate data. It's fairly performant, but having the whole page load when you click on any of the tabs on the left side makes the page feel a little slower and jarring than it needs to be.

And if we need to keep track of what tab is selected in the URL we could use a query param rather than a path param.

ColinBuyck commented on sites/public/src/components/account/ApplicationsView.tsx:
```diff
@@ -0,0 +1,230 @@
+      applicationsService
+        .list({ userId: profile.id })
+        .then((apps) => {
+          apps?.items?.length > 0 ? setApplications(apps.items) : setLoading(false)
```
This was carried over from the previous implementation and controls the loading overlay. I think the idea is that if there are no applications, we don't need to wait for the listings fetch to occur and we can replace the overlay with the message but if there are applications we should keep the loading overlay until the listings fetch completes.

ColinBuyck commented:
Thanks for the review @emilyjablonski ! I've fixed most of these but just have a couple notes. I changed the throttle envs to match hba prod but would like to bring it up to the team on Monday to discuss what value seems right. I'm a little worried about undermining security (if the throttle limit extends to prod) based on this one feature. Secondly, I explored the view thing we discussed and should've noted it in the description but my understanding is that we're using views to make complex queries (ie. joins) easier rather than get a subset of fields in a single table. For example, "fundamentals" view still returns all the fields below
<img width="687" alt="Screenshot 2024-08-08 at 2 32 04 PM" src="https://github.com/user-attachments/assets/c55c3707-b4a2-4443-8dcd-8ab950e6887b">

Might still be worth it to do a custom query like getPublicUserEmailInfo() if we think the number of fields is driving the slowdown. Alternatively if its the number of calls on the db we could batch the ids and then make a single call. Orrr explore using context

Finally, do you mind providing the steps to repdouce the "See all applications" state. That should read as "See listings" but I am also noticing the lack of numbers showing so I'm wondering if its an edge case caused by the throttling

ColinBuyck commented:
@emilyjablonski Also updating to 1 review needed since some of this logic is altered for the next lottery FE ticket.

emilyjablonski reviewed (approved):
🥣

ludtkemorgan commented on sites/public/src/components/account/ApplicationsView.tsx:
```diff
@@ -0,0 +1,229 @@
+        let open = 0,
+          closed = 0,
+          lottery = 0
+        res.forEach((app) => {
+          if (app.fullListing?.status === ListingsStatusEnum.active) {
+            open++
+            if (props.filterType === ApplicationsFilterEnum.Open) displayApplications.push(app)
+          } else if (app.fullListing?.lotteryStatus === LotteryStatusEnum.publishedToPublic) {
+            lottery++
+            if (props.filterType === ApplicationsFilterEnum.Lottery) displayApplications.push(app)
+          } else {
+            closed++
+            if (props.filterType === ApplicationsFilterEnum.Closed) displayApplications.push(app)
+          }
+        })
```
I don't know if this is the spot that is causing the issue. But the placement of applications seems to be incorrect. For example in this screenshot the "Accepting applications" shows a count of 1 but the application is showing up in the "Applications closed"

![image](https://github.com/user-attachments/assets/49aec7f0-8c06-4122-bafa-2fadc3725dac)

ludtkemorgan replied:
This has been fixed

ludtkemorgan reviewed (approved):
Thanks for making those changes. Colin and I talked about this offline, but the throttling issue is due to the large number of api calls. It is recommended to create a new backend endpoint that returns the appropriate data rather than calling it once for every application and every listing. This will be addressed at a later point
